### PR TITLE
VPR PvPMainComboFeature fix (incl some 7.05 fixes)

### PR DESCRIPTION
**Pull request by aldros-ffxi**
Noticed an edge case that inspired me to rework a bit of the logic for the PvP combo. Need to test it tomorrow, but there's a degenerate edge case where your Venom buff can fall off after you hit step 3, but before you press the step 3 button, meaning it'd default back to HindsbaneFang. The problem there is that it could happen when you're in Flank* mode, meaning the button's unusable.

This logic rework fixes that, and should be slightly more performant as well since it reduces the number of calls to OriginalHook().

### CODE REVIEW

File: XIVComboExpanded/Combos/VPR.cs [hunk 1]
```diff
@@ -466,51 +466,60 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
                     return VPR.ThirdGeneration;
                 if (gauge.AnguineTribute == maxtribute - 3)
                     return VPR.FourthGeneration;
+                // Handle a case where you might spam the key and end up missing Ouroboros. I believe PvP handles this case the same way.
+                // Debatable whether we should just return Ouroboros here as a convenience, but the spirit of it was to make it function like PvP which doesn't.
+                return VPR.SteelFangs;
```
[the remark this answers is not in the document]
Comment (aldros-ffxi) [reply XIVComboExpanded/Combos/VPR.cs:471]: Yeah, I think you're right lol

File: XIVComboExpanded/Combos/VPR.cs [hunk 2]
```diff
@@ -466,51 +466,59 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
...
+            // Switch case here for optimization, rather than calling OriginalHook in a lot of places.
+            switch (OriginalHook(VPR.SteelFangs))
             {
-                var noxious = FindTargetEffect(VPR.Debuffs.NoxiousGash);
-                if (level >= VPR.Levels.DreadFangs && (noxious is null || noxious?.RemainingTime < 12)) // 12s hopefully means we won't miss anything on a Reawaken window
-                    return VPR.DreadFangs;
+                // Combo step 1, detect presence of debuff
+                case VPR.SteelFangs:
+                    var noxious = FindTargetEffect(VPR.Debuffs.NoxiousGash);
+                    if (level >= VPR.Levels.DreadFangs && (noxious is null || noxious?.RemainingTime < 13)) // 13s hopefully means we won't miss anything on a Reawaken window
```
Comment (kaedys) [inline XIVComboExpanded/Combos/VPR.cs:479]: The full Reawaken combo takes a minimum of 13.2s before skill speed (2.2s for Reawaken, 2.0s for each of the Generations, and 3.0s for Ouroboros).  If this case is being hit, then the next action will be the start of a new combo, which means after this is used, a minimum of an additional 7.5s (before speed) is needed _on top of that 13.2s_ before Dread Fangs can be used again.  Meaning if the buff is at less than 20.7s and we use Steel Fangs here, and then the Reawaken combo, you'll drop the debuff before getting back to Dread Fangs (I actually suspect this is why they are increasing at least the duration cap on Noxious, it's far too easy to drop the debuff during the burst phase).

Despite it potentially being less than optimal, I would actually set this to 20s.  Regardless, though, I'd argue that it needs to be a _minimum_ of 14-15s to ensure the debuff doesn't get dropped before Ouroboros.

Comment (aldros-ffxi) [reply XIVComboExpanded/Combos/VPR.cs:479]: That's the exact math I was looking for! xD

Comment (MKhayle) [reply XIVComboExpanded/Combos/VPR.cs:479]: 🌔

**Comment by MKhayle** [comment]
(tell me when you're done with this PR)

**Comment by kaedys** [comment]
May also be worth checking out [this thread](https://github.com/MKhayle/XIVComboExpanded/discussions/338#discussioncomment-10169516).  **vitharr137** and **Grammernatzi** came up with a way to check if an action ID is _usable_, ie. if it is greyed out or not.  I've confirmed at least that it works for Tsubame, which is not enabled by a buff or anything else, so it should also work for stuff like the Coils/Dens and similar, and any other ability that us enabled by some prior action but _not_ predicated on a buff.  The actual code content is in [this post](https://github.com/MKhayle/XIVComboExpanded/discussions/338#discussioncomment-10168999) (though you also have to add a definition and instantiation of the `clientStructActionManager` field for it to work).

File: XIVComboExpanded/Combos/VPR.cs [hunk 3]
```diff
@@ -540,44 +548,44 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
...
+                    var noxious = FindTargetEffect(VPR.Debuffs.NoxiousGash); // TODO: Would be useful to handle the case with no target
+                    if (level >= VPR.Levels.DreadMaw && (noxious is null || noxious?.RemainingTime <= 20)) // 20s determined by https://github.com/MKhayle/XIVComboExpanded/pull/353/files#r1690057020
+                        return VPR.DreadMaw;
+
+                    return VPR.SteelMaw;
+
+                case VPR.HuntersBite:
+                    var swift = FindEffect(VPR.Buffs.Swiftscaled);
+                    var instinct = FindEffect(VPR.Buffs.HuntersInstinct);
+                    if (swift is null) // I think we'd always want to prioritize swift since it speeds up the rotation
```
Comment (kaedys) [inline XIVComboExpanded/Combos/VPR.cs:567]: You know, I was going to contest this, but it turns out that it's correct.  Samurai uses Gekko before Kasha, so I assumed the same priority would hold for Viper, damage buff before haste buff, but the official Balance opener is Swiftskin's Bite -> Dreadwinder -> Hunter's Coil, rather than the opposite.

Comment (MKhayle) [reply XIVComboExpanded/Combos/VPR.cs:567]: I am so glad we're getting a division between nerd combos and basic combos

**Comment by MKhayle** [comment]
this will probably need to be revised because of 7.05 patch notes and VPR adjustments (bye bye Noxious Gash)
I will also probably temporarily disable the PvP Main Combo & PvP Main AoE Combo which make use of it in their logic for now.

**Comment by MKhayle** [comment]
don't hate on me and hate SE instead please
also you're 100% allowed to force push your changes in the PR and ignore the conflicts

File: XIVComboExpanded/Combos/VPR.cs (hunk 2 above)
Comment (kaedys) [reply XIVComboExpanded/Combos/VPR.cs]: #whelp

**Comment by aldros-ffxi** [comment]
Yep, changes weren't too bad to incorporate, no hate at all here :)

**Comment by aldros-ffxi** [comment]
According to @kaedys it seems there's probably a bigger rework in the future

**Comment by aldros-ffxi** [comment]
For now I'll hold a few here, since I use the PvP combos, but I'll look at rebasing with @kaedys 's rework logic

**Comment by kaedys** [comment]
Rewrite complete, #368.  Shouldn't be _too_ hard to re-integrate the PvP combos, just slap them in in the correct place inside the `ViperFangs` and `ViperMaws` classes.  Just make sure you place them such that other combos that should take priority are evaluated first.

**Comment by MKhayle** [comment]
my brain doesn't understand what you're doing with those 3 pull requests at the same time 
gib instructions

**Comment by aldros-ffxi** [comment]
Based on the existence of the rewrite, seems like this one is worth closing

**Comment by kaedys** [comment]
You probably could have kept it open, but you'll need to rewrite the PvP combos anyway, for the changes in this PR, the changes in 7.05 (which you'd already included), and the changes to the class structure in that file.  It _shouldn't_ be too tough to do, but lemme know.

**Comment by aldros-ffxi** [comment]
Yeah, I'll just use your commit as a base and open a new one
